pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go

```diff
@@ -38,3 +75,240 @@ func alterTableAddConstraint(
+func getNonDropResultColumns(b BuildCtx, tableID catid.DescID) (ret colinfo.ResultColumns) {
+	for _, colID := range getNonDropColumnIDs(b, tableID) {
+		colElem := mustRetrieveColumnElem(b, tableID, colID)
+		ret = append(ret, colinfo.ResultColumn{
+			Name:           mustRetrieveColumnNameElem(b, tableID, colID).Name,
+			Typ:            mustRetrieveColumnTypeElem(b, tableID, colID).Type,
+			Hidden:         colElem.IsHidden,
+			TableID:        tableID,
+			PGAttributeNum: uint32(colElem.PgAttributeNum),
+		})
+	}
+
+	return ret
+}
+
+// columnLookupFn can look up information of a column by name.
+// It should be exclusively used in DequalifyAndValidateExprImpl.
+func columnLookupFn(
+	b BuildCtx, tableID catid.DescID, columnName tree.Name,
+) (exists bool, accessible bool, id catid.ColumnID, typ *types.T) {
+	columnID := getColumnIDFromColumnName(b, tableID, columnName)
+	if columnID == 0 {
+		return false, false, 0, nil
+	}
+
+	colElem := mustRetrieveColumnElem(b, tableID, columnID)
+	colTypeElem := mustRetrieveColumnTypeElem(b, tableID, columnID)
+	return true, !colElem.IsInaccessible, columnID, colTypeElem.Type
+}
+
+// generateUniqueCheckConstraintName generates a unique name for check constraint.
+// The default name is `check_` followed by all columns in the expression (e.g. `check_a_b_a`).
+// If that name is already in use, append a number (i.e. 1, 2, ...) to it until
+// it's not in use.
+func generateUniqueCheckConstraintName(b BuildCtx, tableID catid.DescID, ckExpr tree.Expr) string {
+	name := checkConstraintDefaultName(b, tableID, ckExpr)
+
+	if constraintNameInUse(b, tableID, name) {
+		i := 1
+		for {
+			numberedName := fmt.Sprintf("%s%d", name, i)
+			if !constraintNameInUse(b, tableID, numberedName) {
+				name = numberedName
+				break
+			}
+			i++
+		}
+	}
+
+	return name
+}
+
+// constraintNameInUse returns whether `name` has been used by any other non-dropping constraint.
+func constraintNameInUse(b BuildCtx, tableID catid.DescID, name string) (ret bool) {
+	scpb.ForEachConstraintName(b.QueryByID(tableID).Filter(publicTargetFilter), func(
+		_ scpb.Status, _ scpb.TargetStatus, e *scpb.ConstraintName,
+	) {
+		if e.Name == name {
+			ret = true
+		}
+	})
+	return ret
+}
+
+// DefaultName creates a check constraint name based on the columns referenced
+// in the check expression. The format is "check_col1_col2...". If columns are
+// duplicated in the expression, they will be duplicated in the name.
+//
+// For example:
+//
+//	CHECK (a < 0) => check_a
+//	CHECK (a < 0 AND b = 'foo') => check_a_b
+//	CHECK (a < 0 AND b = 'foo' AND a < 10) => check_a_b_a
+//
+// Note that the generated name is not guaranteed to be unique among the other
+// constraints of the table.
+func checkConstraintDefaultName(b BuildCtx, tableID catid.DescID, expr tree.Expr) string {
+	var nameBuf bytes.Buffer
+	nameBuf.WriteString("check")
+
+	iterateColNamesInExpr(b, tableID, expr, func(columnName string) error {
```

Comment:
this is nice and I like it. but do we need to consider postgres compatibility for this? (maybe not)

Reply:
I don't think so bc this is what the legacy schema changer did.